Add new gall guide. (#1249)
urbit/urbit.org: content/docs/userspace/gall-guide/2-agent.md

+++
title = "2. The Agent Core"
weight = 10
template = "doc.html"
+++

In this lesson we'll look at the basic type and structure of a Gall agent.

> A Gall agent is a `door` with exactly ten arms. Each arm is responsible for

Reviewer: ```suggestion
A Gall agent is a [door](/docs/glossary/door) with exactly ten [arms](/docs/glossary/arm). Each arm is responsible for
```

> handling certain kinds of events that Gall feeds in to the agent. A `door` is

Reviewer: ```suggestion
handling certain kinds of events that Gall feeds in to the agent. A door is
```

> just a `core` with a sample - it's made with the

Reviewer: ```suggestion
just a [core](/docs/glossary/core) with a sample - it's made with the
```

[barcab](/docs/hoon/reference/rune/bar#_-barcab) rune (`|_`) instead of the
usual [barcen](/docs/hoon/reference/rune/bar#-barcen) rune (`|%`).

## The ten arms

We'll discuss each of the arms in detail later. For now, here's a quick summary.
The arms of an agent can be be roughly grouped by purpose:

### State management

These arms are primarily for initializing and upgrading an agent.

- `on-init`: Handles starting an agent for the first time.
- `on-save`: Handles exporting an agent's state - typically as part of the
  upgrade process but also when suspending, uninstalling and debugging.
- `on-load`: Handles loading a previously exported agent state - typically as
  part of the upgrade process but also when resuming or reinstalling an agent.

### Request handlers

These arms handle requests initiated by outside entities, e.g. other agents,
HTTP requests from the front-end, etc.

- `on-poke`: Handles any calls from outside processes that aren't subscriptions,

>   i.e. one-time actions.

Reviewer: Agents can poke themselves, so this needs to be rephrased slightly to account for this.

- `on-watch`: Handles subscription requests from other entities.
- `on-leave`: Handles unsubscribe requests from other, currently subscribed
  entities.

### Response handlers

These two arms handle responses to requests our agent previously initiated.

- `on-agent`: Handles subscription updates and request acknowledgements from
  other agents.
- `on-arvo`: Handles subscription updates, request acknowledgements, and
  responses from vanes.

### Scry handler

- `on-peek`: Handles local read-only requests.

### Failure handler

- `on-fail`: Handles certain kinds of crash reports from Gall. Most agents don't
  make use of this arm and just leave it with a basic default.

## Bowl

> The sample of the `door` is a `bowl:gall`. Every time an event triggers the

Reviewer: ```suggestion
The sample of a Gall agent door is always a `bowl:gall`. Every time an event triggers the
```

agent, Gall populates the bowl with things like the current date-time, entropy,
subscription information, which ship the request came from, etc, so that all the
arms of the agent have access to that data. For the exact structure and contents
of the bowl, have a read through [its entry in the Gall vane types
documentation](/docs/arvo/gall/data-types#bowl).

## State

If you've worked through [Hoon School](/docs/hoon/hoon-school/intro), you may
recall that a core is a cell of `[battery payload]`. The battery is the core

> itself compiled to nock, and the payload is the subject which it operates on.

Reviewer: ```suggestion
itself compiled to Nock, and the payload is the subject which it operates on.
```

> For an agent, the payload will at least contain the bowl, the usual standard

Reviewer: ```suggestion
For an agent, the payload will at least contain the bowl, the usual Hoon and `zuse` standard
```

3. That arm processes the event, returning a list `card`s to be sent off, and
   the agent core itself with a modified state in the payload.
4. Gall sends the `card`s off and saves the modified agent core.
5. Rinse and repeat.

## Virtualization

When a crash occurs in the kernel, the system usually aborts the computation and
discards the event as though it never happened. Gall on the other hand
virtualizes all its agents, so this doesn't happen. Instead, when a crash occurs
in an agent, Gall intercepts the crash and takes appropriate action depending on
the kind of event that caused it. For example, if a poke from another ship
caused a crash in the `on-poke` arm, Gall will respond to the poke with a
"nack", a negative acknowledgement, telling the original ship the poke was
rejected.

What this means is that you can intentionally design your agent to crash in
cases it can't handle. For example, if a poke comes in with an unexpected
`mark`, it crashes. If a permission check fails, it crashes. This is quite
different to most programs written in procedural languages, which must handle
all exceptions to avoid crashing.

## Example

Here's about the simplest valid Gall agent:

```hoon

> |_  =bowl:gall

Reviewer: I initially suggested adding `^-  agent:gall` to the top of this, but it alters the pretty printer output for `skeleton` below in a way that makes it slightly more obtuse (like it doesn't show the face for `bowl`):
```
10|gqq
  [ [ [our=@p src=@p dap=@tas]
      [   wex
        nlr(
          [p=[wire=/ ship=@p term=@tas] q=[acked=?(%.y %.n) path=/]]
        )
        sup=nlr([p=it(/) q=[p=@p q=/]])
      ]
      act=@ud
      eny=@uvJ
      now=@da
      byk=[p=@p q=@tas r=?([%da p=@da] [%tas p=@tas] [%ud p=@ud])]
    ]
    < 7.ouc
      14.zun
      33.ehb
      14.dyd
      53.vlb
      77.lrt
      232.oiq
      51.qbt
      123.zao
      46.hgz
      1.pnw
      %140
```
That's a bit annoying. We should encourage people to cast, but not when it makes the pretty printer output slightly less clear. I guess its fine to leave it out for now as long as the rest of the examples do it.

Maybe we can remark that typically you'd always cast, but we don't here for clarity.

++  on-init   `..on-init
++  on-save   !>(~)
++  on-load   |=(vase `..on-init)
++  on-poke   |=(cage !!)
++  on-watch  |=(path !!)
++  on-leave  |=(path `..on-init)
++  on-peek   |=(path ~)
++  on-agent  |=([wire sign:agent:gall] !!)
++  on-arvo   |=([wire sign-arvo] !!)
++  on-fail   |=([term tang] `..on-init)
--
```

This is just a dummy agent that does absolutely nothing - it has no state and
rejects all messages by crashing. We'll get to what each of the arms do later.
For now, we'll just consider a few particular points.

> Firstly, note its structure - it's a `door` (created with `|_`) with a sample of

Reviewer: ```suggestion
Firstly, note its structure - it's a door (created with `|_`) with a sample of
```

`|mount %base`. On the Unix side, navigate to `/path/to/fake/ship/base`, and save
the above agent in the `/app` directory as `skeleton.hoon`. Back in the dojo,
commit the file to the desk with `|commit %base`.

For the moment we won't install our `%skeleton` agent. Instead, we'll use the
`%build-file` thread to build it and save it in the dojo's subject so we can
have a look. Run the following in the dojo:

```
> =skeleton -build-file %/app/skeleton/hoon
```

Now, let's have a look:

```
> skeleton
< 10.fxw
  [   bowl
    [ [our=@p src=@p dap=@tas]
      [ wex=nlr([p=[wire=/ ship=@p term=@tas] q=[acked=?(%.y %.n) path=/]])
        sup=nlr([p=it(/) q=[p=@p q=/]])
      ]
      act=@ud
      eny=@uvJ
      now=@da
      byk=[p=@p q=@tas r=?([%da p=@da] [%tas p=@tas] [%ud p=@ud])]
    ]
    <17.zbp 33.wxp 14.dyd 53.vlb 77.wir 232.wfe 51.qbt 123.zao 46.hgz 1.pnw %140>
  ]
>
```

The dojo pretty-prints cores with a format of `number-of-arms.hash`. You can see
the head of `skeleton` is `10.fxw` - that's the battery of the core, our 10-arm
agent. If we try printing the head of `skeleton` we'll see it's a whole lot of

> compiled `nock`:

Reviewer: ```suggestion
compiled Nock:
```

]
```

Currently `skeleton` has no state, but of course in practice you'd want to store
some actual data. We'll add `foo=42` as our state with the `=+` rune at the
beginning of our agent:

```hoon
=+  foo=42
|_  =bowl:gall
++  on-init   `..on-init
++  on-save   !>(~)
++  on-load   |=(vase `..on-init)
++  on-poke   |=(cage !!)
++  on-watch  |=(path !!)
++  on-leave  |=(path `..on-init)
++  on-peek   |=(path ~)
++  on-agent  |=([wire sign:agent:gall] !!)
++  on-arvo   |=([wire sign-arvo] !!)
++  on-fail   |=([term tang] `..on-init)
--
```

Save the modified `skeleton.hoon` in `/app` on the `%base` desk like before, and run `|commit %base` again in the dojo. Then, rebuild it with the same `%build-file` command as before:

```
> =skeleton -build-file %/app/skeleton/hoon
```

If we again examine our agent core's payload by looking at the tail of
`skeleton`, we'll see `foo=42` is now included:

```
> +.skeleton
[   bowl
  [ [our=~zod src=~zod dap=%$]
    [wex={} sup={}]
    act=0
    eny=0v0
    now=~2000.1.1
    byk=[p=~zod q=%$ r=[%ud p=0]]
  ]
  foo=42
  <17.zbp 33.wxp 14.dyd 53.vlb 77.wir 232.wfe 51.qbt 123.zao 46.hgz 1.pnw %140>
]
```

## Summary

> - A Gall agent is a `door` with ten arms and a sample of `bowl:gall`.

Reviewer: ```suggestion
- A Gall agent is a door with exactly ten specific arms and a sample of `bowl:gall`.
```

A Gall agent is a [door](/docs/glossary/door) with exactly ten [arms](/docs/glossary/arm). Each arm is responsible for
handling certain kinds of events that Gall feeds in to the agent. A door is
just a [core](/docs/glossary/core) with a sample - it's made with the
[barcab](/docs/hoon/reference/rune/bar#_-barcab) rune (`|_`) instead of the
usual [barcen](/docs/hoon/reference/rune/bar#-barcen) rune (`|%`).

## The ten arms

We'll discuss each of the arms in detail later. For now, here's a quick summary.
The arms of an agent can be be roughly grouped by purpose:

### State management

These arms are primarily for initializing and upgrading an agent.

- `on-init`: Handles starting an agent for the first time.
- `on-save`: Handles exporting an agent's state - typically as part of the
  upgrade process but also when suspending, uninstalling and debugging.
- `on-load`: Handles loading a previously exported agent state - typically as
  part of the upgrade process but also when resuming or reinstalling an agent.

### Request handlers

These arms handle requests initiated by outside entities, e.g. other agents,
HTTP requests from the front-end, etc.

> - `on-poke`: Handles any calls that aren't subscriptions, i.e. one-time actions.

Reviewer 2: > any calls that aren't subscriptions

Why not just say "handles one-off events" or, indeed, "one-time actions"? Scries aren't subscriptions, but they don't go into `+on-poke`. etc etc

- `on-watch`: Handles subscription requests from other entities.

> - `on-leave`: Handles unsubscribe requests from other, currently subscribed

Reviewer 2: ```suggestion
- `on-leave`: Handles unsubscribe notifications from other, previously subscribed
```

Or maybe "signals" in place of "notifications"?

  entities.

### Response handlers

These two arms handle responses to requests our agent previously initiated.

> - `on-agent`: Handles subscription updates and request acknowledgements from

Reviewer 2: Nitpicky, but I would put these in their chronological order:

```suggestion
- `on-agent`: Handles request acknowledgements and subscription updates from
```

  other agents.

> - `on-arvo`: Handles subscription updates, request acknowledgements, and

Reviewer 2: > subscription updates

Technically yes, things like clay `%many` and dill `%view`s behave like subscriptions, but they're not subscriptions in the same way agent-to-agent `%watch`es are subscriptions. Consider using different phrasing here to avoid conflating the two.

  responses from vanes.

### Scry handler

- `on-peek`: Handles local read-only requests.

### Failure handler

- `on-fail`: Handles certain kinds of crash reports from Gall. Most agents don't

>   make use of this arm and just leave it with a basic default.

Reviewer 2: I would either remove the "most agents don't use it" qualification here, or go on the specify what one _might_ use it for. (That is, recovering from failure to process gracefully. ie if processing a timer failed, I might still want to set a follow-up timer, if we need to "keep going" or w/e.)

## Bowl

The sample of a Gall agent door is always a `bowl:gall`. Every time an event triggers the

> agent, Gall populates the bowl with things like the current date-time, entropy,

Reviewer 2: ```suggestion
agent, Gall populates the bowl with things like the current date-time, fresh entropy,
```

all exceptions to avoid crashing.

## Example

Here's about the simplest valid Gall agent:

```hoon
|_  =bowl:gall
++  on-init   `..on-init
++  on-save   !>(~)
++  on-load   |=(vase `..on-init)
++  on-poke   |=(cage !!)
++  on-watch  |=(path !!)
++  on-leave  |=(path `..on-init)
++  on-peek   |=(path ~)
++  on-agent  |=([wire sign:agent:gall] !!)
++  on-arvo   |=([wire sign-arvo] !!)
++  on-fail   |=([term tang] `..on-init)
--
```

This is just a dummy agent that does absolutely nothing - it has no state and
rejects all messages by crashing. Typically we'd cast this to an `agent:gall`,
but in this instance we won't so it's easier to examine its structure in the
dojo. We'll get to what each of the arms do later. For now, we'll just consider
a few particular points.

Firstly, note its structure - it's a door (created with `|_`) with a sample of
`bowl:gall` and the ten arms described earlier.

Secondly, you'll notice some of the arms return:

```hoon
`..on-init
```

A backtick at the beginning is an irregular syntax meaning "prepend with null",
so for example, in the dojo:

```
> `50
[~ 50]
```

The next part has a dot-dot-arm syntax, which means "the parent core of the

> specified arm". Since we've specified the arm as `on-init`, its parent is our

Reviewer 2: May or may not be worth detailing how this is just `a.b` where `a` happens to be "the subject" aka `.`. iow, it's not special syntax.

the tail is just all the usual standard library functions:

```
> +.skeleton
[   bowl
  [ [our=~zod src=~zod dap=%$]
    [wex={} sup={}]
    act=0
    eny=0v0
    now=~2000.1.1
    byk=[p=~zod q=%$ r=[%ud p=0]]
  ]
  <17.zbp 33.wxp 14.dyd 53.vlb 77.wir 232.wfe 51.qbt 123.zao 46.hgz 1.pnw %140>
]
```

Currently `skeleton` has no state, but of course in practice you'd want to store
some actual data. We'll add `foo=42` as our state with the `=+` rune at the
beginning of our agent:

```hoon
=+  foo=42
|_  =bowl:gall
++  on-init   `..on-init
++  on-save   !>(~)
++  on-load   |=(vase `..on-init)
++  on-poke   |=(cage !!)
++  on-watch  |=(path !!)
++  on-leave  |=(path `..on-init)
++  on-peek   |=(path ~)
++  on-agent  |=([wire sign:agent:gall] !!)
++  on-arvo   |=([wire sign-arvo] !!)
++  on-fail   |=([term tang] `..on-init)
--
```

Save the modified `skeleton.hoon` in `/app` on the `%base` desk like before, and run `|commit %base` again in the dojo. Then, rebuild it with the same `%build-file` command as before:

```
> =skeleton -build-file %/app/skeleton/hoon
```

If we again examine our agent core's payload by looking at the tail of

> `skeleton`, we'll see `foo=42` is now included:

Reviewer 2: This is a great illustration of how state get included with the agent, nice work!

```
> +.skeleton
[   bowl
  [ [our=~zod src=~zod dap=%$]
    [wex={} sup={}]
    act=0
    eny=0v0
    now=~2000.1.1
    byk=[p=~zod q=%$ r=[%ud p=0]]
  ]
  foo=42
  <17.zbp 33.wxp 14.dyd 53.vlb 77.wir 232.wfe 51.qbt 123.zao 46.hgz 1.pnw %140>
]
```

## Summary

- A Gall agent is a door with exactly ten specific arms and a sample of `bowl:gall`.
- Each of the ten arms handle different kinds of events - Gall calls the
  appropriate arm for the kind event it receives.
- The ten arms fit roughly into five categories:
  - State management.
  - Request handlers.
  - Response handlers.
  - Scry handler.
  - Failure handler.

> - The state of an agent—the data it's storing—lives in the core's payload.

Reviewer 2: ```suggestion
- The state of an agent — the data it's storing — lives in the core's payload.
```

Or maybe this isn't typographically correct? Feels more legible to me this way, but up to you.
